_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+While the documentation shows how to use this bijector, the step from theoretical understanding to coding a "black box" may seem wide. If you're anything like the author, here you might feel the urge to "look under the hood" and verify that things really are the way you're assuming. So let's give in to curiosity and allow ourselves a little escapade into the source code.
+
+Peeking ahead, this is how we'll construct a masked autoregressive flow in TFP (again using the still new-ish R bindings provided by [tfprobability](http://github.io/tfprobability)):
+
+```{r}
+library(tfprobability)
+
+maf <- tfb_masked_autoregressive_flow(
+    shift_and_log_scale_fn = tfb_masked_autoregressive_default_template(
+      hidden_layers = list(num_hidden, num_hidden),
+      activation = tf$nn$tanh)
+)
+```
+
+
+Pulling apart the relevant entities here, `tfb_masked_autoregressive_flow` is a bijector, with the usual methods `tfb_forward()`, `tfb_inverse()`, `tfb_forward_log_det_jacobian()` and `tfb_inverse_log_det_jacobian()`.
+The default `shift_and_log_scale_fn`, `tfb_masked_autoregressive_default_template`, constructs a little neural network of its own, with a configurable number of hidden units per layer, a configurable activation function and optionally, other configurable parameters to be passed to the underlying `dense` layers. It's these dense layers that have to respect the autoregressive property. Can we take a look at how this is done? Yes we can, provided we're not afraid of a little Python.
+
+`masked_autoregressive_default_template` (now leaving out the `tfb_` as we've entered Python-land) uses `masked_dense` to do what you'd suppose a thus-named function might be doing: construct a dense layer that has part of the weight matrix masked out. How? We'll see after a few Python setup statements.
+
+```{python}
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+tfd = tfp.distributions
+tfb = tfp.bijectors
+tf.enable_eager_execution()
+```
+
+
+The following code snippets are taken from `masked_dense` (in its [current form on master](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/bijectors/masked_autoregressive.py)), and when possible, simplified for better readability, accommodating just the specifics of the chosen example - a toy matrix of shape 2x3:
+
+```{python}
+# construct some toy input data (this line obviously not from the original code)
+inputs = tf.constant(np.arange(1.,7), shape = (2, 3))
+
+# (partly) determine shape of mask from shape of input
+input_depth = tf.compat.dimension_value(inputs.shape.with_rank_at_least(1)[-1])
+num_blocks = input_depth
+num_blocks # 3
+```
+
+Our toy layer should have 4 units:
+
+```{python}
+units = 4
+```
+
+The mask is initialized to all zeros. Considering it will be used to elementwise multiply the weight matrix, we're a bit surprised at its shape (shouldn't it be the other way round?). No worries; all will turn out correct in the end. 
```

Comment:
So why didn't we start with the right shape?

Reply:
I think they probably do it because it makes `gen_slices` easier to write...? But my perspective here is just the reverse engineer's, I'm just telling my story of stepping through the code...